Add integration for Vevor BLE Heaters

iotmaestro opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->

None


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This PR adds an integration for Vevor BLE heaters (e.g. https://www.aliexpress.com/item/1005005853546232.html)


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: N/A
- This PR is related to issue: N/A
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/29475

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

iotmaestro committed:
Add initial implementation of Vevor Heater integration.

iotmaestro committed:
Update the title generator

iotmaestro committed:
Add unit tests for Vevor BLE heater integration.

iotmaestro committed:
Improve discovery flow for Vevor BLE devices.

iotmaestro committed:
Split the coordinator to a separate file.

iotmaestro committed:
Propagate device name properly

bdraco commented on tests/components/vevor_heater/test_sensor.py:
```diff
@@ -0,0 +1,212 @@
+                error=HeaterError.NO_ERROR,
+            )
+
+            await hass.data[DOMAIN][entry.entry_id].async_refresh()
```
Please don't access hass.data in new tests. Instead use the async_fire_timed_changed test helper

iotmaestro replied:
Done.

bdraco replied:
It looks like this change was missed

iotmaestro replied:
Removed.

bdraco commented on tests/components/vevor_heater/test_sensor.py:
```diff
@@ -0,0 +1,212 @@
+                error=HeaterError.NO_ERROR,
+            )
+
+            await hass.data[DOMAIN][entry.entry_id].async_refresh()
```
As above

iotmaestro replied:
Done.

bdraco commented on tests/components/vevor_heater/test_sensor.py:
```diff
@@ -0,0 +1,212 @@
+            assert await hass.config_entries.async_unload(entry.entry_id)
+            await hass.async_block_till_done()
+
+            pass
```
Useless pass?

iotmaestro replied:
Removed.

bdraco commented on tests/components/vevor_heater/test_sensor.py:
```diff
@@ -0,0 +1,212 @@
+            await hass.async_block_till_done()
+            status_to_set = VevorHeaterStatus(power_status=PowerStatus.OFF)
+
+            await hass.data[DOMAIN][entry.entry_id].async_refresh()
```
As above

iotmaestro replied:
Done.

bdraco commented on tests/components/vevor_heater/test_config_flow.py:
```diff
@@ -0,0 +1,113 @@
+    """Test we successfully create the device if we can't connect to it."""
+
+    with patch(
+        "homeassistant.components.vevor_heater.config_flow.validate_device",
```
Please patch the pypi code and avoid patching the integration (except for async_setup_entry)

iotmaestro replied:
Done.

bdraco commented on tests/components/vevor_heater/test_config_flow.py:
```diff
@@ -0,0 +1,113 @@
+    """Test we successfully create the device if it has a name."""
+
+    with patch(
+        "homeassistant.components.vevor_heater.config_flow.validate_device",
```
Please patch the pypi code and avoid patching the integration (except for async_setup_entry)

bdraco commented on tests/components/vevor_heater/test_config_flow.py:
```diff
@@ -0,0 +1,113 @@
+    """Test we successfully create the device if it doesn't have a name."""
+
+    with patch(
+        "homeassistant.components.vevor_heater.config_flow.validate_device",
```
Please patch the pypi code and avoid patching the integration (except for async_setup_entry)

bdraco commented on tests/components/vevor_heater/test_sensor.py:
```diff
@@ -0,0 +1,212 @@
+    status_to_set = VevorHeaterStatus(power_status=PowerStatus.OFF)
+
+    with patch(
+        "homeassistant.components.bluetooth.async_ble_device_from_address",
```
Ok to patch the Bluetooth public api

iotmaestro replied:
Acknowledged.

bdraco commented on homeassistant/components/vevor_heater/sensor.py:
```diff
@@ -0,0 +1,143 @@
+
+    entities = []
+    _LOGGER.debug("got sensors: %s", coordinator.data)
+    entities.extend(
```
Looks like there is no need to make a list and a generator expression can be passed to async_add_entities

iotmaestro replied:
I *definitely* didn't steal this from another integration without thinking twice about it :laughing: 

Done.

bdraco replied:
Doesn't seem updated, did you forget to push?

iotmaestro replied:
Discussed elsewhere, resolving.

bdraco commented on homeassistant/components/vevor_heater/coordinator.py:
```diff
@@ -0,0 +1,66 @@
+        self._ble_device = device
+        self._vevor_device = VevorDevice(
+            address=self._ble_device.address,
+            name="Vevor " + (self._ble_device.name or ""),
```
Please use f-strings

iotmaestro replied:
Done.

bdraco commented on homeassistant/components/vevor_heater/config_flow.py:
```diff
@@ -0,0 +1,123 @@
+        """Confirm discovery."""
+        assert self._discovery_info is not None
+        discovery_info = self._discovery_info
+        title = "Vevor " + str(discovery_info.name or discovery_info.address)
```
Please use f-strings

iotmaestro replied:
Done.

bdraco commented on homeassistant/components/vevor_heater/config_flow.py:
```diff
@@ -0,0 +1,123 @@
+
+async def validate_device(hass: HomeAssistant, address: str) -> dict[str, Any]:
+    """Validate the given device."""
+    if not address:
```
This looks unreachable through the normal flow

iotmaestro replied:
AIUI and as my experiments show this gets called -  at the BT discovery step (`async_step_bluetooth`). Given the entire integration is reverse engineered and there's no documentation for the device I think actually communicating with the device like we do here is the only way to find out if it is what we think it is. 

I'm of course happy to take advice on the matter as my experience with HA design principles and best practices is... limited.

bdraco replied:
address is typed to `str` so if this was reachable it would be a bug in the bluetooth stack. We should take it out

bdraco commented on homeassistant/components/vevor_heater/__init__.py:
```diff
@@ -0,0 +1,66 @@
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Vevor Heater from a config entry."""
+
+    hass.data.setdefault(DOMAIN, {})
```
Please don't set anything in hass.data until after passing the first refresh successfully

iotmaestro replied:
Done.

bdraco commented on homeassistant/components/vevor_heater/sensor.py:
```diff
@@ -0,0 +1,143 @@
+class VevorHeaterSensor(CoordinatorEntity[VevorHeaterUpdateCoordinator], SensorEntity):
+    """Vevor BLE heater sensors for the device."""
+
+    def __init__(
```
Please see

https://developers.home-assistant.io/docs/core/entity/#has_entity_name-true-mandatory-for-new-integrations

iotmaestro replied:
The entire naming section is a great piece of documentation I was looking for and didn't find. Can we reference it from somewhere underneath the "Integrations 101" section?

bdraco commented on homeassistant/components/vevor_heater/sensor.py:
```diff
@@ -0,0 +1,143 @@
+        device_class=SensorDeviceClass.ENUM,
+    ),
+    "input_voltage": SensorEntityDescription(
+        key="input_voltage",
```
Please add entity_category so diagnostic entities don't appear on default dashboard

iotmaestro replied:
Done - can you please double check the classification makes sense? The ones I labeled as "config" can technically be updated but the integration currently doesn't support it.

bdraco replied:
Sensor entities should never be config since they can't actually interact with device beyond providing data. Diagnostic is likely better for this case

bdraco commented on homeassistant/components/vevor_heater/const.py:
```diff
@@ -0,0 +1,8 @@
+
+DOMAIN = "vevor_heater"
+
+CHAR_UUID_HEATER_CONTROL = UUID("0000ffe1-0000-1000-8000-00805f9b34fb")
```
Please move specific uuids to the pypi code and import instead

iotmaestro replied:
It's not used in HA Python code (leftover from having everything in the HA repo) so just removed it. I assume it's ok to keep it in the json manifest for discovery purposes.

bdraco replied:
Yes definitely fine in the manifest.  In general, anything that's protocol details / magic numbers for the device should be in the pypi code, but that doesn't apply to the manifest

bdraco replied:
It looks like it didn't actually get removed

iotmaestro replied:
It should be now, sorry!

bdraco commented on homeassistant/components/vevor_heater/coordinator.py:
```diff
@@ -0,0 +1,66 @@
+        self._ble_device = device
+        self._vevor_device = VevorDevice(
+            address=self._ble_device.address,
+            name="Vevor " + (self._ble_device.name or ""),
```
Looks like this isn't using f-strings

iotmaestro replied:
Different local workflows apparently. I tick off all items I've addressed locally and consider PRs in draft mode as a scratch space which should only be reviewed when explicitly marked as such. Apologies for wasting your time if you're used to a different paradigm. Let me finish this round of comments my way and I'll behave better next time :).

bdraco replied:
Usually I check the review comment against the code before marking it resolved. But it's fine to do it now you are doing it now that I understand your workflow

iotmaestro replied:
Ack - will be more careful with assumptions next time when contributing to a new project.

iotmaestro committed:
Addressed PR feedback comments.

iotmaestro committed:
Mark all sensors as diagnostic for the time being.

iotmaestro commented:
All comments should be addressed, please take another look.

iotmaestro commented:
@bdraco friendly ping

bdraco commented:
@iotmaestro Please resolve conflicts.

Sorry, I'm really behind on reviews as I'm traveling a lot this month.  Hoping to have more time when we get into December, but maybe before than.

bdraco commented on homeassistant/components/vevor_heater/__init__.py:
```diff
@@ -0,0 +1,66 @@
+    vevor_coordinator = VevorHeaterUpdateCoordinator(hass=hass, device=ble_device)
+
+    await vevor_coordinator.async_config_entry_first_refresh()
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = vevor_coordinator
```
```suggestion
    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = vevor_coordinator
```

bdraco replied:
Looks like this was missed

iotmaestro replied:
Done.

bdraco commented on homeassistant/components/vevor_heater/__init__.py:
```diff
@@ -0,0 +1,66 @@
+    ) -> None:
+        vevor_coordinator.set_ble_device(service_info.device)
+
+    bluetooth.async_register_callback(
+        hass,
+        _async_update_ble,
+        BluetoothCallbackMatcher(address=address),
+        BluetoothScanningMode.ACTIVE,
+    )
```
The callback needs to be cancelled when the integration is unloaded
Wrap in 

`entry.async_on_unload`

iotmaestro replied:
Done.

bdraco commented on homeassistant/components/vevor_heater/__init__.py:
```diff
@@ -0,0 +1,66 @@
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+    if unload_ok and DOMAIN in hass.data:
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    return unload_ok
```
```suggestion
    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
        hass.data[DOMAIN].pop(entry.entry_id)

    return unload_ok
```

unload will never be called unless its loaded

iotmaestro replied:
Done.

bdraco commented on homeassistant/components/vevor_heater/coordinator.py:
```diff
@@ -0,0 +1,66 @@
+        self._ble_device = device
+        self._vevor_device = VevorDevice(
+            address=self._ble_device.address,
+            name=f"Vevor {self._ble_device.name or ''}",
```
maybe use short address from `bluetooth_data_tools` if the name is unavailable

iotmaestro replied:
Done

bdraco commented on homeassistant/components/vevor_heater/sensor.py:
```diff
@@ -0,0 +1,157 @@
+    coordinator: VevorHeaterUpdateCoordinator = hass.data[DOMAIN][entry.entry_id]
+    _LOGGER.debug("got sensors: %s", coordinator.data)
+
+    async_add_entities(
+        [VevorHeaterSensor(coordinator, sensor) for sensor in SENSORS_MAPPINGS.values()]
+    )
```
This can be a generator expression

iotmaestro replied:
Done

bdraco commented on homeassistant/components/vevor_heater/sensor.py:
```diff
@@ -0,0 +1,157 @@
+        self.entity_description = entity_description
+
+        self._attr_unique_id = (
+            f"{coordinator.get_device_address()}_{entity_description.key}"
```
AFAICT address shouldn't ever change so lets make it a property instead


Also its called twice in this function. Store it in a local var instead of calling it twice

iotmaestro replied:
`get_device_address` is a simple getter so there's no performance implication to calling it multiple times, but extracted to a separate variable to make this unquestionable.

The address is not used anywhere else so I don't think it needs to be a property.

bdraco commented on homeassistant/components/vevor_heater/sensor.py:
```diff
@@ -0,0 +1,157 @@
+    def _propagate_value(self) -> None:
+        """Update attrs from device."""
+        if self.coordinator.data.status is not None:
+            self._attr_native_value = getattr(
+                self.coordinator.data.status, self.entity_description.key
+            )
```
Instead of this use a small function for each description to get the value so it can be typed

bdraco replied:
That works.  I think the current practice is to call it `value_fn`

iotmaestro replied:
Perfect, renamed.

bdraco commented on tests/components/vevor_heater/test_config_flow.py:
```diff
@@ -0,0 +1,159 @@
+
+    with patch(
+        "homeassistant.components.bluetooth.async_ble_device_from_address",
+        return_value=BLEDevice(
+            address=service_info.address,
+            name=service_info.name,
+            details=defaultdict(str),
+            rssi=FIXTURE_SERVICE_INFO.rssi,
+        ),
```
There is a helper for constructing BLEDevices in the bluetooth integration tests

iotmaestro replied:
Done.

bdraco commented on tests/components/vevor_heater/test_sensor.py:
```diff
@@ -0,0 +1,225 @@
+    status_to_set = VevorHeaterStatus(power_status=PowerStatus.OFF)
+
+    with patch(
+        "homeassistant.components.bluetooth.async_ble_device_from_address",
```
This would be better a a pytest.fixture since you need to patch it for the duration of the test, and it will save a lot of indent

iotmaestro replied:
I'm having issues with the interactions between multiple fixtures doing this (patch can also be used as a function decorator, bu that does weird things to the hass parameter). I tried untangling it for a while and then gave up - as this part is significant to what the test does, it may be better to keep it explicit anyway.

bdraco replied:
Can you move it into the patch block below to save some indent and make it a bit more readable instead?

iotmaestro replied:
Done.

bdraco reviewed (changes requested):
Please see comments above

iotmaestro commented:
> Please see comments above

Thank you for the comments. This time round it's me traveling so please bear with me addressing the comments, thank you!

iotmaestro committed:
Addressed PR comments.

iotmaestro committed:
Fix merge issues with codeowners

bdraco commented on homeassistant/components/vevor_heater/coordinator.py:
```diff
@@ -0,0 +1,67 @@
+        self._ble_device = device
+        self._vevor_device = VevorDevice(
+            address=self._ble_device.address,
+            name=f"Vevor {self._ble_device.name or bluetooth_data_tools.short_address(self._ble_device.address)}",
```
It would be nicer to make `short_address` a local var.  I try not to use function calls in `f-strings` as it makes it harder to read it takes longer to review.  Keep in mind we have to check every function for blocking I/O in the event loop, and while I know this one, its still a lot of time spent to do that.

iotmaestro replied:
Makes sense - extracted to a separate local var but kept the or clause as to avoid the extra work when we end up not needing the short name.

While here I also gave address the same treatment.

bdraco commented on homeassistant/components/vevor_heater/sensor.py:
```diff
@@ -0,0 +1,186 @@
+T = TypeVar("T")
+
+
+@dataclass
```
https://developers.home-assistant.io/blog/2023/12/11/entity-description-changes/

iotmaestro replied:
Oh perfect, there goes my thoughtful comment :).

bdraco commented on tests/components/vevor_heater/test_sensor.py:
```diff
@@ -0,0 +1,223 @@
+    """Tests sensors when the heater is set to target temperature heating."""
+    status_to_set = VevorHeaterStatus(power_status=PowerStatus.OFF)
+
+    with patch(
+        "homeassistant.components.bluetooth.async_ble_device_from_address",
+        return_value=FIXTURE_BLE_DEVICE,
```
as above

iotmaestro replied:
Done.

bdraco commented on tests/components/vevor_heater/test_sensor.py:
```diff
@@ -0,0 +1,223 @@
+    """Tests sensors when the heater is off."""
+    status_to_set = VevorHeaterStatus(power_status=PowerStatus.OFF)
+
+    with patch(
+        "homeassistant.components.bluetooth.async_ble_device_from_address",
+        return_value=FIXTURE_BLE_DEVICE,
+    ):
```
as above

iotmaestro replied:
Done.

bdraco commented on tests/components/vevor_heater/test_sensor.py:
```diff
@@ -0,0 +1,223 @@
+            await hass.async_block_till_done()
+
+
+def _assert_state(
+    state: Any, expected_value: Any, expected_unit: Optional[str]
+) -> None:
+    if expected_value is None:
+        # If we expect the state not to be present it can't have any unit.
+        # Failing assert indicates a misconfiguration of the test, not a SUT malfunction.
+        assert expected_unit is None
+
+        # Now check the SUT
+        assert state is None
+    else:
+        assert state.state == expected_value
+        if expected_unit is None:
+            assert "unit" not in state.attributes
+        else:
+            assert state.attributes["unit_of_measurement"] == expected_unit
```
We don't want branching in tests as it makes them harder to maintain, review, and follow.

There are usually two options to avoid this:

1. Break the test up into two tests
2. use `@pytest.mark.parametrize` (lots of examples in the code base)

iotmaestro replied:
This is not a branching test per se, it's just a helper method to evaluate the presence / absence of a state according to the expectation. Similar logic can be found e.g. in assert methods of `unittest.testcase`.

bdraco commented on homeassistant/components/vevor_heater/config_flow.py:
```diff
@@ -0,0 +1,120 @@
+                await validate_device(self.hass, address)
+                self._discovered_devices[address] = discovery_info.name
+            except CannotConnect:
+                return self.async_abort(reason="not_supported")
```
`not_supported` doesn't seem like the right error for a connection problem

iotmaestro replied:
Right. CannotConnect conflated two error conditions - that we can't establish a BT connection, and that there was an error using the protocol we implement.

I split the error into two with appropriate handling.

bdraco reviewed (changes requested):
Looks close to ready. 👍  I added a few more comments above.

bdraco commented:
Please don't hard pin deps in the lib. Instead use `>=`

`    vevor-heater-ble 0.0.5 depends on bleak-retry-connector~=3.2.1`

iotmaestro committed:
vevor_heater: Address more PR comments.

iotmaestro commented:
> Please don't hard pin deps in the lib. Instead use `>=`
> 
> ` vevor-heater-ble 0.0.5 depends on bleak-retry-connector~=3.2.1`

I'm a little confused about what this refers to. The manifest specifies an exact version which seems to be common practice looking at other integrations, and the BLE library itself uses fuzzy version matching (https://github.com/iotmaestro/vevor-heater-ble/blob/main/pyproject.toml). Can you elaborate?

iotmaestro committed:
vevor: Rename extractor to value_fn to follow convention.

iotmaestro committed:
Merge remote-tracking branch 'upstream/dev' into vevor-init

bdraco commented:
The manifest is the only place we want the specific version pins.  The libraries themselves should specify only a minimum version with optionally a maximum next major version (but we try to avoid this). If libraries constrain to only specific versions, than HA can never be installed because we would have hundreds of dependency conflicts which could never be solved.

iotmaestro committed:
vevor_heater: Fix test failures

iotmaestro committed:
vevor: Update the BLE dependency to the latest version.

This is to cherry pick
https://github.com/iotmaestro/vevor-heater-ble/commit/c546d9afea77b6fe27cdcb3f357013a4ca542119.

iotmaestro commented:
> The manifest is the only place we want the specific version pins. The libraries themselves should specify only a minimum version with optionally a maximum next major version (but we try to avoid this). If libraries constrain to only specific versions, than HA can never be installed because we would have hundreds of dependency conflicts which could never be solved.

I see - I read the comment the other way around (that `~=3.2.1` is the desired state). Something something 5AM wakeups. Fixed now, and also the tests. Please take another look.

Note for self: Once the PR looks ok let's not forget about updating the "introduced in" versions in this and all related PRs.

iotmaestro committed:
Merge remote-tracking branch 'upstream/dev' into vevor-init

bdraco reviewed (changes requested):
CI is failing, please check. Thanks

iotmaestro committed:
vevor: Bump BLE dependency version to fix the package name.

bdraco commented on homeassistant/components/vevor_heater/coordinator.py:
```diff
@@ -0,0 +1,69 @@
+        """Attempt to update the sensor data."""
+        try:
+            await self._vevor_device.refresh_status(self._ble_device)
+        except Exception as err:
```
Please only trap expected exceptions here. We want unknown or unexpected ones to still raise

bdraco commented on tests/components/vevor_heater/test_sensor.py:
```diff
@@ -0,0 +1,211 @@
+        assert await hass.config_entries.async_unload(entry.entry_id)
+        await hass.async_block_till_done()
+
+        pass
```
This looks unnecessary

bdraco commented on tests/components/vevor_heater/test_sensor.py:
```diff
@@ -0,0 +1,211 @@
+        assert await hass.config_entries.async_unload(entry.entry_id)
+        await hass.async_block_till_done()
+
+        pass
```
This looks unnecessary

bdraco reviewed (changes requested):
Please add coverage for missing lines or add the files to `.coveragerc` (except config_flow.py which requires 100% coverage before merging)
